### ISSUE TITLE
chromedp Version 2

### ISSUE THREAD

**Issue by pmurley** [chromedp/chromedp#280]
First of all, thanks to the team who works on this. It's already a great tool and I've gotten a lot out of it.

I'm seen an upcoming chromedp v2 mentioned in some issues, and I'm wondering if you have a rough estimate what time timeline looks like for a release.

My use case involves handling lots of DevTools events ( e.g. #252 ) and is also heavily dependent on timeouts ( #246 ) for API calls, both of which I've seen mentioned as features which will be improved significantly in v2. I'm trying to decide if I should wait a bit to see if v2 comes out, or press forward with the current version.

Again, thanks for your time and your great work so far!

**Comment by mvdan** [comment]
You can see the current progress in the [refactor](https://github.com/chromedp/chromedp/tree/refactor) branch. The API is already quite different, and the tests pass, but the internal design issues haven't been fixed yet.

For timeout and racy issues, the temporary fix we're using at our startup is short sleeps and polling. I realise that's not ideal, but if you can live with code that's temporarily slower, that's an option. There's no way to fix this better in master without the v2 refactor.

Properly handling events is in a similar spot. Right now, you could fork chromedp and add callbacks or channels to get notified of events, which might be enough for your use case. But properly exposing all events can only really happen in v2.

As for the timeline, I imagine we could have an alpha with the big issues fixed (timeouts and events) in a matter of weeks. Having a finished release with all the features we want, good test coverage, few bugs, and updated examples is likely going to take twice as much time.

Long story short, if you can wait around a month, I'd wait. Otherwise, you should be able to get stuff done with a bit of hacking, like we do.

**Comment by mvdan** [comment]
I'll assume that my answer above resolved your question :)

**Comment by pmurley** [comment]
Yup, my fault. Thanks!

**Comment by m4p1e** [comment]
i am waiting for it :)

**Comment by mvdan** [comment]
Please see https://github.com/chromedp/chromedp/issues/168#issuecomment-475672811 - if anyone can help testing the refactor branch, that would be helpful.
